src/type-data-frame.c

```diff
@@ -0,0 +1,57 @@
+#include "vctrs.h"
+#include "utils.h"
+
+
+// [[ include("utils.h") ]]
+SEXP new_data_frame(SEXP x, R_len_t n) {
+  x = PROTECT(r_maybe_duplicate(x));
+
+  init_data_frame(x, n);
+
+  if (Rf_length(x) == 0) {
+    Rf_setAttrib(x, R_NamesSymbol, vctrs_shared_empty_chr);
+  }
+
+  UNPROTECT(1);
+  return x;
+}
+
+// [[ include("utils.h") ]]
+bool is_compact_rownames(SEXP x) {
+  return Rf_length(x) == 2 && INTEGER(x)[0] == NA_INTEGER;
+}
+// [[ include("utils.h") ]]
+R_len_t compact_rownames_length(SEXP x) {
+  return abs(INTEGER(x)[1]);
+}
+
+static void init_compact_rownames(SEXP x, R_len_t n);
+static SEXP new_compact_rownames(R_len_t n);
+
+// [[ include("utils.h") ]]
+void init_data_frame(SEXP x, R_len_t n) {
+  Rf_setAttrib(x, R_ClassSymbol, classes_data_frame);
+  init_compact_rownames(x, n);
+}
+// [[ include("utils.h") ]]
+void init_tibble(SEXP x, R_len_t n) {
+  Rf_setAttrib(x, R_ClassSymbol, classes_tibble);
+  init_compact_rownames(x, n);
+}
+
+static void init_compact_rownames(SEXP x, R_len_t n) {
+  SEXP rn = PROTECT(new_compact_rownames(n));
+  Rf_setAttrib(x, R_RowNamesSymbol, rn);
+  UNPROTECT(1);
+}
+static SEXP new_compact_rownames(R_len_t n) {
+  if (n <= 0) {
```

Comment:
This is mildly surprising — would it be better to always use the same representation of row names?

Reply:
This is from `.set_row_names()`'s implementation. It's probably safer to reproduce its behaviour exactly.